src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/PkgControl.java

```diff
@@ -40,28 +47,130 @@
     private final PkgControl parent;
     /** The full package name for the node. */
     private final String fullPackage;
+    /**
+     * The regex pattern for partial match (exact and for subpackages) - only not
+     * null if regex is true.
+     */
+    private final Pattern patternForPartialMatch;
+    /** The regex pattern for exact matches - only not null if regex is true. */
+    private final Pattern patternForExactMatch;
+    /** If this package represents a regular expression. */
+    private final boolean regex;
 
     /**
      * Construct a root node.
      * @param pkgName the name of the package.
+     * @param regex flags interpretation of pkgName as regex pattern.
      */
-    PkgControl(final String pkgName) {
+    PkgControl(final String pkgName, final boolean regex) {
         parent = null;
+        this.regex = regex;
         fullPackage = pkgName;
+        if (regex) {
+            patternForPartialMatch = createPatternForPartialMatch();
+            patternForExactMatch = createPatternForExactMatch();
+        }
+        else {
+            patternForPartialMatch = null;
+            patternForExactMatch = null;
+        }
     }
 
     /**
-     * Construct a child node.
+     * Construct a child node. The concatenation of regular expressions needs special care:
+     * see {@link #ensureSelfContainedRegex(String, boolean)} for more details.
      * @param parent the parent node.
      * @param subPkg the sub package name.
+     * @param regex flags interpretation of subPkg as regex pattern.
      */
-    PkgControl(final PkgControl parent, final String subPkg) {
+    PkgControl(final PkgControl parent, final String subPkg, final boolean regex) {
         this.parent = parent;
-        fullPackage = parent.fullPackage + "." + subPkg;
+        if (regex || parent.regex) {
```

Comment:
I still can not clearly read a logic behind grouping and forcing to regexp, smth is not smooth .
Base on previous discussion that we have to force plain-text patter from parent to regexp if child is regexp, .... looks like we have to whole below branch of PkgConrolls to be regexp if some PkgControll is regexed.

why we care only about current and parent regex state ?
What about case : 
```
<import-control pkg="com.puppycrawl.tools.checkstyle" regex="true">
    <subpackage name="checks" >
       <subpackage name="imports" >
```

Looks like import "import" Pkgcontrol will not be forced to regexp.
Please point me to Uts that cover it, I did not find it , and explain a bit logic if I am wrong again.